File: Daily/sbgo/Leetcode628-三个数的最大乘积.go
package sbgo

import (
	"sort"
	"math"
)

func maximumProduct(nums []int) int {
	sort.Slice(nums,func(i,j int) bool {return nums[i] < nums[j]})
	n := len(nums)
	return max(nums[0] * nums[1] * nums[n-1],nums[n-3]*nums[n-2]*nums[n-1])
}


func maximumProduct2(nums []int) int {
	min1,min2 := math.MaxInt64,math.MaxInt64
	max1,max2,max3 := math.MinInt64,math.MinInt64,math.MinInt64

	for _,num := range nums{
		if num < min1{
			min2 = min1
			min1 = num
		}else if num < min2 {
			min2 = num
		}

		if num > max1{
			max3 = max2 
			max2 = max1
			max1 = num
		}else if num > max2{
			max3 = max2
			max2 = num
		}else if num > max3{
			max3 = num
		}
	}
	return max(max1 * max2 * max3,min1 * min2 * max1)
}

File: Leetcode/PY/Leetcode628-三个数的最大乘积.py
from typing import List
class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        nums.sort()
        n = len(nums)
        return max(nums[0] * nums[1] * nums[n-1],nums[n-3] * nums[n - 2] * nums[n - 1])

class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        #max1最大
        max1,max2,max3 = -float('inf'),-float('inf'),-float('inf')
        #min1最小
        min1,min2 = float('inf'),float('inf')
        for num in nums:
            if num < min1:
                min2 = min1
                min1 = num
            elif num < min2:
                min2 = num
        
            if num > max1:
                max3 = max2
                max2 = max1
                max1 = num
            elif num > max2:
                max3 = max2
                max2 = num
            elif num > max3:
                max3 = num
        return max(min1 * min2 * max1,max1 * max2 * max3)